### PR TITLE
nixos/grub: Add submenu for each generation with specialisation

### LINKED ISSUE #247507: Documentation: Specilisation Old Generation Boot Entries
## Problem

NixOS specialisations are supposed to have their own entries in the Grub menu for previous generations according to the [nixos wiki page](https://nixos.wiki/wiki/Specialisation#Boot_entries). However in practice that is not the case. The older generations do not hold any boot entry for the specialisations, but only the default entry.

nixpkgs: `nixos-23.05`

## Proposal

I am not sure, whether to update the wiki to correctly mention current behavior, or whether to fix the grub module to generate entries for specialisations for older generations.

## Checklist

- [x] checked [latest Nixpkgs manual] \([source][nixpkgs-source]) and [latest NixOS manual] \([source][nixos-source])
- [x] checked [open documentation issues] for possible duplicates
- [x] checked [open documentation pull requests] for possible solutions

### PR DESCRIPTION
**Pull request by dev-null-undefined**
Before this PR there was no way to access (boot into) specialisation of previous generations from grub, even tho they are there.

This PR will add grub submenu for each generation if the generation has any specialisation. Which will allow you to boot into them.

Example with 2 current specialisation, generation 4 having 2 specialisation and generation 3 having only 1.

Grub menu before:
```
NixOS - Default
NixOS - (Specialisation 1 - date - version)
NixOS - (Specialisation 2 - date - version)
NixOS - All configurations:
+  NixOS - Configuration 4 (date - version)
+  NixOS - Configuration 3 (date - version)
+  NixOS - Configuration 2 (date - version)
+  NixOS - Configuration 1 (date - version)
```

Now:
```

NixOS - Default
NixOS - (Specialisation 1 - date - version)
NixOS - (Specialisation 2 - date - version)
NixOS - All configurations:
+  > NixOS - Configuration 4 (date - version)
+  +  NixOS - Configuration 4 - Default (date - version)
+  +  NixOS - Configuration 4 (Specialisation 1 - date - version)
+  +  NixOS - Configuration 4 (Specialisation 2 - date - version)
+  > NixOS - Configuration 3 (date - version)
+  +  NixOS - Configuration 3 - Default (date - version)
+  +  NixOS - Configuration 3 (Specialisation 1 - date - version)
+  NixOS - Configuration 2
+  NixOS - Configuration 1
```

Note: symbol `+` in the example means that the entry is inside submenu that is the first one above with one less `+`
symbol `>` is used for the user to indicate that the entry is submenu.

Fixes #247507

### CODE REVIEW

**Comment by dev-null-undefined** [comment]
If anyone want's to test this out you can do so by running
```bash
# Build the boot loader installer based on your current config
# which is needed because of device IDs and other config
nix build /etc/nixos#nixosConfigurations.$(hostname).config.system.build.installBootLoader --override-input nixpkgs github:dev-null-undefined/nixpkgs/grub-specialisations --refresh
# Then I would recommend BACKING UP your `/boot`!!
# nixos is very good at saving you but not if you destroy your boot which this possibly could do

# And run the installer script
sudo ./result /nix/var/nix/profiles/system
```

File: nixos/modules/system/boot/loader/grub/install-grub.pl [hunk 1]
```diff
@@ -516,38 +516,53 @@ sub addEntry {
...
+    my ($name, $nameSuffix, $path, $options, $current) = @_;
 
-# Add default entries.
-$conf .= "$extraEntries\n" if $extraEntriesBeforeNixOS;
+    # Do not search for grand children
+    my @links = sort (glob "$path/specialisation/*");
 
-addEntry("@distroName@ - Default", $defaultConfig, $entryOptions, 1);
+    if (scalar(@links) != 0) {
+        $conf .= "submenu \"Submenu for $name$nameSuffix\" --class submenu {\n";
```
Comment (samueldr) [inline nixos/modules/system/boot/loader/grub/install-grub.pl:526]: I'm not fond of adding "Submenu for" to every generations. Though I agree there needs to be a way to determine whether or not this is going to boot a system of go into a submenu.

The phrasing could be "Specialisations for ...", but still a bit cumbersome to me.

Maybe a symbol? Unicode would work on default configs, but not if the user uses other fonts. So let's keep it simple :(.

```suggestion
        $conf .= "submenu \"$name$nameSuffix >\" --class submenu {\n";
```

Not sure what else other than a `>` at the end could work here.

**Review by samueldr — COMMENTED** [review]
This is basically approved and almost merged.

Only a nit and a less nitty but resolved ~~problem~~ preference to look into.

File: nixos/modules/system/boot/loader/grub/install-grub.pl (hunk 1 above)
Comment (samueldr) [reply nixos/modules/system/boot/loader/grub/install-grub.pl:526]: There's still the question about how to handle these...

In my opinion the "Submenu for" prefix becomes extra-redundant given that in my experience, it is quite rare to have only a couple generations with specialisations... In my use-case, they are used for a long time.

But I do agree that there needs to be a signal helping the user know "this one is a submenu, this one is not".

Any thoughts?

**Comment by dev-null-undefined** [comment]
@samueldr Updated the PR message and force pushed the suggested changes, thanks for the help <3

**Comment by samueldr** [comment]
Thanks a lot! Procrastinating doing a thing paid once again ;)

**Pull request merged**